### PR TITLE
feat(inverted_index.create): add external sorter

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

* Add the `Sorter` trait for handling sorting of incremental input.
* Add the `ExternalSorter` implementation, together with the dependency `ExternalTempFileProvider` to enable access to external files. The `ExternalSorter` will employ external sorting, has a parameter `memory_usage_threshold`, and tracks memory usage to ensure that the sorting is completed with a minimal memory footprint.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2705

next PR to be proposed:
[feat(inverted_index.create): add index creator](https://github.com/zhongzc/greptimedb/pull/4)

### CODE REVIEW

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            Some(buffer)
+        } else {
+            None
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn shuffle_values_and_sorted_result(
+        row_count: usize,
+    ) -> (Vec<Option<Vec<u8>>>, BTreeMap<Option<Vec<u8>>, Vec<usize>>) {
```
Comment (evenyag) [inline src/index/src/inverted_index/create/sort/external_sort.rs:209]: suggestion(non-blocking): We can add a type alias for `BTreeMap<Option<Vec<u8>>, Vec<usize>>` to help others to understand the contents of the map. e.g. Maybe `ValueRowIds`
```rust
type ValueRowIds = BTreeMap<Option<Vec<u8>>, Vec<usize>>;
```

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+    /// The number of rows per group for bitmap indexing which determines how rows are
+    /// batched for indexing. It is used to determine which segment a row belongs to.
+    segment_row_count: usize,
+
+    /// Tracks memory usage of the buffer
+    current_memory_usage: usize,
+
+    /// The memory usage threshold at which the buffer should be dumped to an external file
+    memory_usage_threshold: usize,
```
Comment (evenyag) [inline src/index/src/inverted_index/create/sort/external_sort.rs:59]: Do we need to define the threshold as `Option<usize>` and use `None` to disable external sort? If we want to disable this by default, the TOML rendered might contain a very large number.

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+/// Sets or appends a bit at the specified offset within the bitmap
+fn set_bit(bitmap: &mut BitVec, offset: usize) {
+    if offset >= bitmap.len() {
+        bitmap.resize(offset + 1, false);
+    }
+    bitmap.set(offset, true);
+}
+
+#[cfg(test)]
```
Comment (evenyag) [inline src/index/src/inverted_index/create/sort/external_sort.rs:180]: Should we add a case for `segment_row_count > 1`?

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    async fn output(&mut self) -> Result<SortOutput> {
+        let readers = self.temp_file_provider.read_all(&self.index_name).await?;
+
+        let buf_values = mem::take(&mut self.values_buffer).into_iter();
+        let mut merging_sorted_stream: SortedStream = Box::new(stream::iter(buf_values.map(Ok)));
+
+        // Sequentially merge each intermediate file's stream into the merged stream
+        for reader in readers {
+            let intermediate = IntermediateReader::new(reader).into_stream().await?;
+            merging_sorted_stream = MergeSortedStream::merge(merging_sorted_stream, intermediate);
```
Comment (evenyag) [inline src/index/src/inverted_index/create/sort/external_sort.rs:90]: non-blocking: Consider using a K-way merge in the future.

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+            .create(&self.index_name, file_id)
+            .await?;
+        IntermediateWriter::new(writer).write_all(values).await?;
+
+        self.current_memory_usage = 0;
+        Ok(())
+    }
+
+    /// Determines the segment index range for the row index range
+    /// [self.total_row_count, self.total_row_count + n - 1]
```
Comment (evenyag) [inline src/index/src/inverted_index/create/sort/external_sort.rs:195]:
```suggestion
    /// `[self.total_row_count, self.total_row_count + n - 1]`
```

**Review by evenyag — APPROVED** [review]

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+impl Sorter for ExternalSorter {
+    /// Pushes a value into the sorter, adding it to the in-memory buffer and dumping the buffer to
+    /// an external file if necessary
+    async fn push(&mut self, value: Option<BytesRef<'_>>) -> Result<()> {
+        self.push_n(value, 1).await
+    }
+
+    /// Pushes n identical values into the sorter, adding them to the in-memory buffer and dumping
+    /// the buffer to an external file if necessary
+    async fn push_n(&mut self, value: Option<BytesRef<'_>>, n: usize) -> Result<()> {
```
Comment (killme2008) [inline src/index/src/inverted_index/create/sort/external_sort.rs:74]: Looks like the default implementation of `push_n` in trait is unnecessary. Maybe we want user to implement `push_n` and let `push`  be the default implementation.

File: src/index/src/inverted_index/create/sort/external_sort.rs [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+    async fn may_dump_buffer(&mut self, memory_diff: usize) -> Result<()> {
+        self.current_memory_usage += memory_diff;
+        if self.memory_usage_threshold.is_none()
+            || self.current_memory_usage < self.memory_usage_threshold.unwrap()
+        {
+            return Ok(());
+        }
+
+        let values = mem::take(&mut self.values_buffer);
+        let file_id = &format!("{:012}", self.total_row_count);
```
Comment (killme2008) [inline src/index/src/inverted_index/create/sort/external_sort.rs:182]: It's better to add some debug log.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@zhongzc There is a clippy warning.

**Comment by zhongzc** [comment]
> There is a clippy warning.

Due to the toolchain update of the develop branch, fixed now.

**Pull request merged**